src/System.Net.Sockets/src/System/Net/Sockets/IOControlKeepAlive.cs

```diff
@@ -0,0 +1,93 @@
+        {
+            SocketKeepAliveCwt = new ConditionalWeakTable<Socket, IOControlKeepAlive>();
+            UIntSize = Marshal.SizeOf<uint>();
+        }
+
+        public static bool TryGet(Socket socket, SocketOptionLevel optionLevel, SocketOptionName optionName, ref int optionValue)
+        {
+            if (!IsNeeded(optionLevel, optionName))
+                return false;
+                
+            IOControlKeepAlive ioControlKeepAlive;
+            SocketKeepAliveCwt.TryGetValue(socket, out ioControlKeepAlive);
+            optionValue = optionName == SocketOptionName.TcpKeepAliveTime ?
+                ToS(ioControlKeepAlive?.timeMs ?? DefaultTimeMs) :
+                ToS(ioControlKeepAlive?.intervalMs ?? DefaultIntervalMs);
+            return true;
+        }
+
+        public static bool TrySet(Socket socket, SocketOptionLevel optionLevel, SocketOptionName optionName, int optionValue)
+        {
+            if (!IsNeeded(optionLevel, optionName))
+                return false;
+
+            var ioControlKeepAlive = SocketKeepAliveCwt.GetOrCreateValue(socket);
+            if (optionName == SocketOptionName.TcpKeepAliveTime)
+                ioControlKeepAlive.timeMs = ToMs(optionValue);
+            else
+                ioControlKeepAlive.intervalMs = ToMs(optionValue);
+            socket.IOControl(IOControlCode.KeepAliveValues, ioControlKeepAlive.ToByteArray(), null);
+            return true;
+        }
+
+        public IOControlKeepAlive()
+        {
+            timeMs = DefaultTimeMs;
+            intervalMs = DefaultIntervalMs;
+        }
+
+        private static bool IsNeeded(SocketOptionLevel optionLevel, SocketOptionName optionName)
+        {
+            return IsWindowsLessThan10Version1709() &&
+                optionLevel == SocketOptionLevel.Tcp &&
+                (optionName == SocketOptionName.TcpKeepAliveTime || optionName == SocketOptionName.TcpKeepAliveInterval);
+        }
+
+        private static bool IsWindowsLessThan10Version1709()
+        {
+            return RuntimeInformation.IsOSPlatform(OSPlatform.Windows) &&
+                Interop.NtDll.RtlGetVersionEx(out Interop.NtDll.RTL_OSVERSIONINFOEX os) == 0 &&
+                (os.dwMajorVersion < 10 || (os.dwMajorVersion == 10 && os.dwMinorVersion == 0 && os.dwBuildNumber < 16299));
+        }
+
+        private static int ToS(uint milliseconds)
+        {
+            return (int)(milliseconds / 1000u);
+        }
+
+        private static uint ToMs(int seconds)
+        {
+            return (uint)seconds * 1000u;
+        }
+
+        private byte[] ToByteArray()
+        {
+            var keepAliveSettings = new byte[3 * UIntSize];
+            BitConverter.GetBytes(OnOff).CopyTo(keepAliveSettings, 0);
+            BitConverter.GetBytes(timeMs).CopyTo(keepAliveSettings, UIntSize);
+            BitConverter.GetBytes(intervalMs).CopyTo(keepAliveSettings, 2 * UIntSize);
+            return keepAliveSettings;
+        }
```

Comment:
Rather than allocating a new byte[] each time, could we use one stored in a [ThreadStatic]? e.g.
```C#
[ThreadStatic]
private static byte[] s_keepAliveValuesBuffer;
...
byte[] buffer = s_keepAliveValuesBuffer ?? (s_keepAliveValuesBuffer = new byte[3 * sizeof(uint)]);
ioControlKeepAlive.Fill(buffer);
socket.IOControl(IOControlCode.KeepAliveValues, buffer, null);
```
That way, we don't have to allocate a new byte[] every time this is called.

Then `Fill` would look something like:
```C#
private void Fill(byte[] buffer)
{
    Debug.Assert(buffer != null);
    Debug.Assert(buffer.Length == 3 * sizeof(uint));

    BitConverter.TryWriteBytes(buffer.AsSpan(), OnOff);
    BitConverter.TryWriteBytes(buffer.AsSpan(sizeof(uint)), _timeMs);
    BitConverter.TryWriteBytes(buffer.AsSpan(sizeof(uint)*2), _intervalMs);
}
```
to avoid the multiple array allocations associated with BitConverter.GetBytes.